### ISSUE TITLE
Should Exercism encourage the students to write their own tests?

### ISSUE THREAD

**Issue by janluke** [exercism/exercism#5073]
### What I like
I really like the idea of working in my local environment, which is what I do anyway when I practice on other sites like leetcode.

### What I don't like
I don't think that the propsed workflow is didactically good. 

**Rude TL;DR:** Exercism doesn't encourage the learner to write his own tests and propose a questionable "micro-scale TDD" workflow that I think it's detrimental both for the learner and for the promotion of TDD.

If there's something I like about sites for coding interviews and competitive programming is that you must take responsibility for what you submit. Before submitting, you don't have a magical oracle that tells you if your solution is correct; you must write your own tests to ensure that. You don't have the right answer and you if you lazily submit before testing you'll probably get an error. This helps in getting used to uncertainty, in learning the importance of testing and in developing testing abilities.

In contrast, Exercism provides you a full test suite and, apparently, it doesn't encourage you to write your own tests and to use the provided tests only at the end (for final validation). Furthermore, it proposes a micro-scale pseudo-TDD workflow that doesn't make sense to me. I'll try to explain why. 

Exercism problems I've seen so far are descriptions of a single "unit of functionality" (usually, small algorithmic problems). So the test suite itself is a unit test. Individual test cases doesn't encapsulate a feature, they just test your program for one or multiple inputs sampled from a relevant subset of the input space (testing by partitioning). Thus, the workflow you are proposing is equivalent to the following:
- partition the input space (already done for you)
- write a test for a subset in the partition (already done for you)
- write some code that makes the tests written so far pass
- repeat

This strategy is especially bad for algorithmic problems. These cannot generally be solved in this way because a solution that works for a subset of the input space cannot usually be extended or adapted to work for the entire input space. Most of the times, you can't work incrementally at this scale. And certainly you can't "discover" a solution to an algorithmic problem by TDD.

In this kind of problems, i personally think that writing the tests first is not much better than writing them later, but why may it be the case? One could argue that writing tests first forces you to think about all possible cases you need to consider before starting developing a solution; in this way, it helps in developing a better picture of the problem. But note that's the case only if you write **ALL** test cases you can think of before starting coding! It doesn't work if you write a single test case and then try to solve the problem for passing that test case and the previous ones.

### Proposal
Encourage the learner to 
- write his own tests and to try to test against the provided test suite only for validation; 
- explain why testing is important and optionally present the _potential_ advantages of writing the tests first.

(Direct consequence: get rid of the "micro-scale TDD" thing.)

**Comment by sshine** [comment]
> other sites like leetcode

I haven't evaluated this site in particular, so I'll compare Exercism to HackerRank to make a point: Other "learn to code" websites have different merits and flaws, and Exercism stands to benefit from copying the good parts of other sites, *when possible*. So when comparing, please justify the comparison by outlining similarities and differences that play a role (e.g. as below, or see #5033 for a similar comparison).

Here are some differences that make copying of good parts difficult:

When it comes to submitting exercises, HackerRank chooses a stdin/stdout interface which means (1) it does not matter how you structure a solution, as long as it upholds the stdin/stdout interface, (2) the server-side test suite is language-independent, (3) the exercise description is more precise and so works better as a spec (most significantly: concrete language-independent examples of input/output).

Right now the Exercism handout tests work like a spec.

If it were not the tests, the exercise description would work like a spec.

The Exercism exercise description is vague because it addresses multiple similar but not equivalent exercises in different programming languages. So if students write and/or submit their own tests, we cannot easily decide when that is satisfactory (and so we cannot partially automate this process).

If we assume that exercise descriptions will change for the better as part of [the Chicago project](https://exercism.io/blog/redesigning-tracks-in-partnership-with-chicago-university-and-sloan-foundation), even though this work hasn't started yet, let's assume that test files *could* eventually be removed and regard the exercise description as the spec.

> provides you a full test suite and, apparently, it doesn't encourage you to write your own tests

That's very true.

Conjecture: Unless a submission in mentored mode is somehow consistently assessed by the quality of its tests, any additional testing will be neglected. I sense this from seeing how few students address parts of an exercise that say "this part is optional".

If testing is a common theme in each submissions, and mentors are expected to discuss the quality of submitted tests in every exercise, I believe Exercism would quickly devolve into being a website that teaches testing practices rather than general [fluency](https://exercism.io/values).

So perhaps a better way to encourage TDD on Exercism (assuming that we can fully decipher an exercise from its textual description, which is not currently the case), is to design each language track such that some early exercises in some way or another *are about testing* (in that language).

This would also be covered under fluency, since when you adapt to a new language, you have to learn its test libraries and methodologies. E.g. Haskell has QuickCheck/Hedgehog, Erlang has Dialyzer. How does test-file auto-discovery work in language X? Is there an Rspec-style library? A JUnit-style library?

Then again, the line between "fluency" and "proficiency" is blurry to me.

**Comment by janLuke** [comment]
> [about Leetcode] I haven't evaluated this site in particular

Sorry, I realized my first sentence wasn't clear. Leetcode and other sites I had in mind (including HackerRank) are browser-based. When I use those sites I almost never use the browser editor because I much prefer working in my preferred IDE. So I'm forced to copy and paste between the browser and my IDE, which is annoying. I like Exercism because it allows easily me to work locally and to keep a local archive of all problems I solved along with a README.

> If it were not the tests, the exercise description would work like a spec.  The Exercism exercise description is vague because it addresses multiple similar but not equivalent exercises in different programming languages.

I think the description should be the spec. Why do exercises have language dependent details? You could have language-agnostic exercises and still decide what exercises to include in each track.

> Conjecture: Unless a submission in mentored mode is somehow consistently assessed by the quality of its tests, any additional testing will be neglected.

First, I don't think that the student should submit his own tests, unless the track is about testing.

Second, in other sites, writing your tests is kinda "optional" but is motivated to write his tests because the code is evaluated only after submission and submitting buggy code has a "price":
- in interview sites, it equates to telling the interviewer you are sure the code works; if it doesn't, you are not in a good position;
- in competition sites, you get a penalty if you submit wrong code.

In real life, "submitting" equates to pushing code in production. The best way to encourage testing is really just simulating a real-world situation, where submitting/releasing wrong code has a price.

> When it comes to submitting exercises, HackerRank chooses a stdin/stdout interface

That's a smart design choice, which I think all other coding sites adopted, even if most of them don't require the student to explicitly read/write from stdin/stout (they just provide a function signature and do the reading/writing behind the scenes). Ideally, I think Exercism should work like that too. But this would be a huge change, which is why I didn't even try to propose that. What I proposed is just to change the flow you suggest to the student:
- during development use your own tests to ensure the code works; 
- use the provided test suite only for validation, when you are sure the code works.

Then the student can decide if it's worth or not to do that, exactly as in other sites.

**Comment by benrodenhaeuser** [comment]
Just two quick notes: 

- The actual proposal seems to be a lot more narrow than the issues title suggests. So maybe tone down the title a bit? Otherwise people might respond to the title rather than the proposal.
- I tried to find the actual language that suggests the "micro-scale TDD" etc. I remember something of the kind, but couldn't track it down. Does anyone have the text we are talking about here at hand?

**Comment by iHiD** [comment]
> PS: Maybe for the title, something like "Change website copy that explains the workflow suggested to users?" could work?

I don't believe this issue is about changing the copy, I believe it's about changing the flow of the website. My understanding was that OP is suggesting getting rid of tests, and encouraging the student to write and submit their own tests based on the README.md. Am I misunderstanding that, @janLuke?

**Comment by benrodenhaeuser** [comment]
I was going off of this part: 

> What I proposed is just to change the flow you suggest to the student:
> - during development use your own tests to ensure the code works;
> - use the provided test suite only for validation, when you are sure the code works.

But @janLuke is the ultimate judge, of course :-)

**Comment by janLuke** [comment]
@benrodenhaeuser Well, you quoted me :) The title could be: "Encourage the student to write his own tests and revise the suggested TDD flow (in the Java track)".

>I tried to find the actual language that suggests the "micro-scale TDD" etc.

The language is Java but I (mistakenly) assumed it was a general suggestion common to all tracks.
TDD is introduced in the 2nd exercise in the track (https://exercism.io/my/solutions/bba56d2b5c2349cab6651ab555a11576), then the suggested flow is explained:

> The test files in this track contain all the tests your solution should pass to be considered valid. That doesn't immediately seem to be compatible with the cycle described above, in which tests are written one by one. However, the tool that we use to write our tests, JUnit, provides an @Ignore annotation that can be used to temporarily skip an already-written test. Using this annotation, we make sure that the test files we deliver to you satisfy the following rules:

> 
>  1. The first test in any test file is not skipped by default.
>  2. All but the first test in any test file are skipped by default.
> 
> This allows you to simulate the TDD cycle by following these slightly-modified steps:
> 
>  1. Run the tests to verify that at most one test currently fails.
>  2. Update your existing code until all the non-skipped tests pass.
>  3. Clean up your code, making sure that all non-skipped tests continue to pass.
>  4. Remove the topmost @Ignore annotation in the test file.
>  5. Return to step 1 until no tests are skipped and all tests pass!

**Comment by sshine** [comment]
> I like Exercism because it allows easily me to work locally

Having both options available are on the roadmap.

Notice that Exercism caters to multiple demographies.

Some may prefer a web-based IDE as they try out a language.

> I think the description should be the spec. Why do exercises have language dependent details?

As I already said, "exercise description *[...]* addresses multiple similar but not equivalent exercises in different programming languages". The concept of "a common exercise between tracks" is a loose one; the same exercise is specified differently in each track, and we actually aim to do that even more:

If you'll read #5088, with v3 tracks we intend to make "concept exercises" for each language specialized to that particular language, and the common exercises we have today would qualify as what #5088 calls "practice exercises". 

How the common interface between tracks is going to work requires not only a proper design, but also for us to grow Exercism's voluntary cooperative so it can handle the kind of challenges we've faced especially this year, cf. exercism/problem-specifications#1560.

> You could have language-agnostic exercises and still decide what exercises to include in each track.

Yes, we could, and we do. But the similarity between e.g. HackerRank's algorithm-centric exercises will not fit Exercism's goal of obtaining language fluency within dozens of individual languages: The obvious way that an exercise may be modelled in one language may make it useful to teach "default return values", while in another language it would teach "monadic error handling". We care about *how* an exercise is made, what an exercise can teach, not only its implementation is correct.

For that reason, a verbatim translation of the good parts of other language learning sites is futile.

> I don't think that the student should submit his own tests, unless the track is about testing.

(Perhaps you mean "unless the *exercise* is about testing", but either way, you make a point.)

I think this is also @iHiD's opinion.

> in other sites, *[...]* the student is motivated to write his tests because *[...]* submitting buggy code has a "price":
>
> - in interview sites, *[...]* your "interview score" drops;
> - in competition sites, you *[...]* get a penalty *[...]*

So since Exercism is neither an interview site nor a competition site, what would be the purpose of a penalty associated with submitting buggy code? One purpose could be to reduce the overhead of mentors spending time on solutions that don't work. Still, since we allow (and in some cases currently encourage) submitting partial solutions if students are stuck, we should have a principal discussion before introducing such a penalty.

> > HackerRank chooses a stdin/stdout interface
>
> I think Exercism should work like that too. But this would be a huge change, which is why I didn't even try to propose that.

Yes, it would be. In my opinion, it does not align with the goal of obtaining language fluency when every single isolated exercise also contains a stdin/stdout-serialization component. For example, in what way would I convey that the composition of a higher-order module in ML follows [best practice](http://homepages.inf.ed.ac.uk/stg/publications/fengshui.pdf) via a stdin/stdout interface? A stdin/stdout-interface would not guarantee that I even addressed the use of a module in the first place.

With most of Exercism's current efforts of (a) specializing language tracks even further, and (b) giving automated feedback via language-specific analysers, the common interface we're most concerned about right now is actually that of #5079's Representers: How can we share automated feedback from language-specific analysers between tracks that share common language idioms?

> What I proposed is just to change the flow you suggest to the student:
>
> - during development use your own tests to ensure the code works;
> - use the provided test suite only for validation, when you are sure the code works.

That seems like a valuable approach.

We'll need to address this differently for "concept exercises" (for which the exercise description can more easily be exact and work as a specification) and "practice exercises" (for which the exercise description is currently both not very exact and is shared between language tracks).

Thanks for your input so far.

**Comment by tqa236** [comment]
Here's my two cents about the test cases from a student's perspective.

Apart from the superb mentoring, the local coding environment and the test suite are what I like the most about Exercism. When I start to learn a new language, it's impossible for me to know how to write a high-quality test case that's not "print them all to the screen." So providing good test cases is also vital for my learning. 

That said, sometimes I'm quite comfortable with the language and I prefer to go even further and cover more edge cases than the test cases, I should write the test myself. But it's not easy to think about all the edge cases without a way to test our solution.

So, I propose a solution like this. For each exercise (let's focus on the core exercise first), if applicable, we divide the test cases into basic and advanced (or obligatory and optional, or core and edge cases,...) and when downloading an exercise, only basic test cases are provided. About the advanced test case, it should only be available when the mentor downloading the solution of the student. Then, depending on the student's ability, they will suggest covering some or all edge cases. It would be the best if we can run the solution automatically on Exercism side to provide feedback immediately. But whatever it is, my basic idea is to divide a test suite into two parts and provide one to the student.

**Comment by janLuke** [comment]
> So since Exercism is neither an interview site nor a competition site, what would be the purpose of a penalty associated with submitting buggy code?

It wasn't my intention to suggest you should introduce some form of penalty. But I'll anwer your question: the didactical value of the flow it would encourage is already a good reason. 

> For example, in what way would I convey that the composition of a higher-order module in ML follows best practice via a stdin/stdout interface? A stdin/stdout-interface would not guarantee that I even addressed the use of a module in the first place.

Of course, the stdin/stdout approach is useful only for checking programs output. It is useless for "concept exercises". But nothing prevents you to have both language-independent tests (for program output testing) and language-dependent test (for "concept testing"). 

> Thanks for your input so far.

You are welcome. Thank you for your time.

We went off-topic, so I'd propose to stop OT conversations here. About the main points of the issue, I think I've written all I wanted to say.

**Comment by sshine** [comment]
> It wasn't my intention to suggest you should introduce some form of penalty.

Perhaps "penalty" is too negative a word.

It wasn't my intention to suggest that penalties are bad.

If we assume for a second that @tqa236's idea of splitting tests into "handout" and "hidden" is added to the workflow, then we do copy a part of what I like about HackerRank's approach to testing: You are actively encouraged to consider edge cases, since even the "submit iteration, check test output" feedback loop has a time cost to the student.

> the stdin/stdout approach is useful only for checking programs output. It is useless for "concept exercises". But nothing prevents you to have both language-independent tests (for program output testing) and language-dependent test (for "concept testing").

One thing (that I've said previously) prevents us:

We want to actively encourage people into writing solutions a certain way (via structure, manual feedback, automated feedback). The less structure an exercise stub provides, the more feedback is necessary. We want to minimize the necessary amount of feedback without telling the student an optimal solution.

We do that by writing a test suite that imports a function from a library file.

We can add a layer of stdin/stdout stuff to every exercise (and possibly even hide this in a separate file, like the tests live in their own file). But if a stdin/stdout interface does not remove the need for language-specific testing, giving all language tracks a common interface, then what have we gained?

> We went off-topic, so I'd propose to stop OT conversations here. About the main points of the issue, I think I've written all I wanted to say.

Yes, so the test interface is a side-track to the principal question of whether Exercism should provide tests or require the user to write them. Which itself is a different question than "Should Exercism be more like *other website* in various ways?" -- perhaps a variation of @tqa236's idea will work with our current track-specific test interface and the infrastructure currently being built around automated testing.

And thanks again for your valuable input.

**Comment by janLuke** [comment]
>But I believe that testing is not just about correctness: It is also practice to learn some language's test framework.

Three considerations:

1. always providing tests without encouraging the student to start writing his own tests is a good recipe for not learning a test framework. If you believed that only reading is an effective way to learn, you would provide even the problem solution along with the test suite :)

2. many provided test suites I've seen so far are clearly not written to be "exemplary"; for example, I remember a suite where the same 10x10 matrix was copied and pasted into multiple test methods for no reason.

3. learning the basics of a test framework is a simple task that should be considered an integral part of being fluent in a language. Even if the framework is not part of the standard library, there's almost always a standard de facto framework.

4. some sites provide 2 or 3 test cases just to remind you the syntax of the test framework and expect you to write the others.

On-topic questions still to be answered:
1. Does the proposed "simulated TDD flow" make sense? Is it effective? Is it the right way to do TDD for this kind of problems?
2. Should we suggest another flow where the student writes his/her tests and use the provided tests only at the end as all other coding sites (implicitly/explicitly) do?

**Comment by snsvrno** [comment]
I like @tqa236's comments with the given, hidden tests, and support @janLuke's latest considerations,

I think a problem statement and a test file should be given, the tests in the file should be basic and all that is required to move on to the next lesson (or pass that lesson). If you pass all the given tests you get 1 star (or something), and if you pass the hidden tests you get 2 star (or whatever). 

The problem statement should explain what is expected of the function (handle null inputs, error when given certain cases, etc) which describe what the hidden tests are, but the user would need to build these tests to check if they pass the cases. When they submit it runs remotely and grades them on if they pass these tests or not.

Writing tests is something I took too long to do, and I'm not even sure if I'm really that good at them, so I think its a good idea to help educate about tests and writing quality tests.

**Comment by ErikSchierboom** [comment]
Hello 👋

With the launch of Exercism v3, we are **closing all issues in this repository** to help give us a clean slate to detect new problems. If this issue is still relevant to Exercism v3 (e.g. it's a feature that we haven't implemented in v3, or a bug that still exists), please reopen it and we will review it and post an update on it as soon as we get chance.

Thanks for helping make Exercism better, and we hope you enjoy v3 🙂
